pkg/controller/seed-controller-manager/mla/ratelimit_cortex_controller.go

```diff
@@ -0,0 +1,248 @@
+	}
+
+	if !kubernetes.HasFinalizer(mlaAdminSetting, mlaFinalizer) {
+		kubernetes.AddFinalizer(mlaAdminSetting, mlaFinalizer)
+		if err := r.Update(ctx, mlaAdminSetting); err != nil {
+			return reconcile.Result{}, fmt.Errorf("updating finalizers: %w", err)
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+type ratelimitCortexController struct {
+	ctrlruntimeclient.Client
+	mlaNamespace string
+
+	log *zap.SugaredLogger
+}
+
+func newRatelimitCortexController(
+	client ctrlruntimeclient.Client,
+	log *zap.SugaredLogger,
+	mlaNamespace string,
+) *ratelimitCortexController {
+	return &ratelimitCortexController{
+		Client:       client,
+		mlaNamespace: mlaNamespace,
+
+		log: log,
+	}
+}
+
+func (r *ratelimitCortexController) ensureLimits(ctx context.Context, mlaAdminSetting *kubermaticv1.MLAAdminSetting) error {
+	configMap := &corev1.ConfigMap{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: r.mlaNamespace, Name: runtimeConfigMap}, configMap); err != nil {
+		return fmt.Errorf("unable to get cortex runtime config map: %w", err)
+	}
+	config, ok := configMap.Data[runtimeConfigFileName]
+	if !ok {
+		return errors.New("unable to find runtime config file in configmap")
+	}
+	or := &overrides{}
+	if err := yaml.Unmarshal([]byte(config), or); err != nil {
+		return fmt.Errorf("unable to unmarshal runtime config[%s]: %w", config, err)
+	}
+
+	tenantOr := tenantOverride{}
+
+	if mlaAdminSetting.Spec.MonitoringRateLimits != nil {
+		if mlaAdminSetting.Spec.MonitoringRateLimits.IngestionRate > 0 {
+			tenantOr.IngestionRate = &mlaAdminSetting.Spec.MonitoringRateLimits.IngestionRate
+		}
+		if mlaAdminSetting.Spec.MonitoringRateLimits.MaxSeriesPerMetric > 0 {
+			tenantOr.MaxSeriesPerMetric = &mlaAdminSetting.Spec.MonitoringRateLimits.MaxSeriesPerMetric
+		}
+		if mlaAdminSetting.Spec.MonitoringRateLimits.MaxSeriesPerQuery > 0 {
+			tenantOr.MaxSeriesPerQuery = &mlaAdminSetting.Spec.MonitoringRateLimits.MaxSeriesPerQuery
+		}
+		if mlaAdminSetting.Spec.MonitoringRateLimits.MaxSamplesPerQuery > 0 {
+			tenantOr.MaxSamplesPerQuery = &mlaAdminSetting.Spec.MonitoringRateLimits.MaxSamplesPerQuery
+		}
+		if mlaAdminSetting.Spec.MonitoringRateLimits.IngestionBurstSize > 0 {
+			tenantOr.IngestionBurstSize = &mlaAdminSetting.Spec.MonitoringRateLimits.IngestionBurstSize
+		}
+		if mlaAdminSetting.Spec.MonitoringRateLimits.MaxSeriesTotal > 0 {
+			tenantOr.MaxSeriesTotal = &mlaAdminSetting.Spec.MonitoringRateLimits.MaxSeriesTotal
+		}
+	}
+	or.Overrides[mlaAdminSetting.Spec.ClusterName] = tenantOr
+	data, err := yaml.Marshal(or)
+	if err != nil {
+		return fmt.Errorf("unable to marshal runtime config[%+v]: %w", or, err)
+	}
+	configMap.Data[runtimeConfigFileName] = string(data)
+	return r.Update(ctx, configMap)
+}
+
+func (r *ratelimitCortexController) cleanUp(ctx context.Context) error {
+	mlaAdminSettingList := &kubermaticv1.MLAAdminSettingList{}
+	if err := r.List(ctx, mlaAdminSettingList, ctrlruntimeclient.InNamespace(r.mlaNamespace)); err != nil {
```

Comment:
This clean up will not work because `MLAAdminSetting` object is in cluster namespace, not in mlaNamespace

Reply:
done